File: day2/src/main/java/PartTwo.java

import java.util.Arrays;
import java.util.stream.IntStream;

public class PartTwo {
    public static int main(int[][] input) {
        return Arrays.stream(input)
                    .mapToInt(line -> isSafe(line))
                    .sum();
    } 

    private static int isSafe(int[] line) {
        if (isAscending(line) || isDescending(line)) return 1;

        for (int i = 0; i < line.length; i++) {
            int[] left = Arrays.copyOfRange(line, 0, i);
            int[] right = Arrays.copyOfRange(line, i + 1, line.length);

            int[] sub = IntStream.concat(Arrays.stream(left), Arrays.stream(right)).toArray();

            // for (int num : sub) System.out.print(num + " ");
            // System.out.println();


            if (isAscending(sub) || isDescending(sub)) return 1;
        }

        return 0;
    }

    private static boolean isAscending (int[] line) {
        int current = line[0];
        for (int i = 1; i < line.length; i++) {
            if (current >= line[i] || current + 3 < line[i]) {
                return false;
            }
            current = line[i];
        }
        return true;
    }

    private static boolean isDescending (int[] line) {
        int current = line[0];
        for (int i = 1; i < line.length; i++) {
            if (current <= line[i] || current - 3 > line[i]) {
                return false;
            }
            current = line[i];
        }
        return true;
    }
}

File: day9/part_two.py
with open('day9.input', 'r') as file:
    string = [line for line in file][0]

big = []

for a, b, idx in zip(range(0, len(string), 2), range(1, len(string) + 1, 2), range(len(string))):
    for _ in range(int(string[a])):
        big.append(idx)
    if b < len(string):
        for _ in range(int(string[b])):
            big.append('.')

left, right = 0, len(big) - 1
l, r = left, right

while (right > -1):
    while right == '.':
        r -= 1
    while r > -1 and big[right] == big[r]:
        r -= 1
    file = right - r

    left = 0
    while(left < right):
        while left < right and big[left] != '.':
            left += 1
        l = left

        while l < right and big[l] == '.':
            l += 1
        
        if l - left >= file:
            # move the file
            big[left : left + file], big[r + 1 : right + 1] = big[r + 1 : right + 1], big[left : left + file]
            break
        else:
            left = l

    right = r

print(sum([idx * int(num) if num != '.' else 0 for idx, num in enumerate(big)]))